### PR TITLE
Feature/s3 c 2788 get obj retention

### PR DESCRIPTION
**Pull request by dora-korpar**
Implements GET Object Retention API. To be merged after https://github.com/scality/Arsenal/pull/1058 and https://github.com/scality/cloudserver/pull/2551

### CODE REVIEW

**Review by tmacro — APPROVED** [review]

File: tests/unit/api/objectGetRetention.js [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    before(() => cleanup());
+
+    describe('without Object Lock enabled on bucket', () => {
+        beforeEach(done => {
+            bucketPut(authInfo, bucketPutRequest, log, err => {
+                assert.ifError(err);
+                objectPut(authInfo, putObjectRequest, undefined, log, done);
+            });
+        });
+        afterEach(() => cleanup());
```
Comment (miniscruff) [inline tests/unit/api/objectGetRetention.js]:
```suggestion
        afterEach(cleanup);
```

File: tests/unit/api/objectGetRetention.js [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        const bucketObjLockRequest = Object.assign({}, bucketPutRequest,
+            { headers: { 'x-amz-bucket-object-lock-enabled': true } });
+
+        beforeEach(done => {
+            bucketPut(authInfo, bucketObjLockRequest, log, err => {
+                assert.ifError(err);
+                objectPut(authInfo, putObjectRequest, undefined, log, done);
+            });
+        });
+        afterEach(() => cleanup());
```
Comment (miniscruff) [inline tests/unit/api/objectGetRetention.js]:
```suggestion
        afterEach(cleanup);
```

File: lib/api/objectGetRetention.js [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+                }
+                if (objectMD.isDeleteMarker) {
+                    if (reqVersionId) {
+                        log.trace('requested version is delete marker',
+                            { method: 'objectGetRetention' });
+                        return next(errors.MethodNotAllowed);
+                    }
+                    log.trace('most recent version is delete marker',
+                        { method: 'objectGetRetention' });
+                    return next(errors.NoSuchKey);
```
Comment (miniscruff) [inline lib/api/objectGetRetention.js:67]: As far as I can tell reading the tests and this code, this is all untested.

File: lib/api/objectGetRetention.js [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+    const { bucketName, objectKey } = request;
+
+    const decodedVidResult = decodeVersionId(request.query);
+    if (decodedVidResult instanceof Error) {
+        log.trace('invalid versionId query', {
+            versionId: request.query.versionId,
+            error: decodedVidResult,
+        });
+        return process.nextTick(() => callback(decodedVidResult));
```
Comment (miniscruff) [inline lib/api/objectGetRetention.js:32]: I believe this path is also untested

**Review by miniscruff — CHANGES_REQUESTED** [review]

File: lib/api/objectGetRetention.js (hunk 3 above)
Comment (dora-korpar) [reply lib/api/objectGetRetention.js:67]: Tests are in a followup PR: https://github.com/scality/cloudserver/pull/2609

File: lib/api/objectGetRetention.js (hunk 4 above)
Comment (dora-korpar) [reply lib/api/objectGetRetention.js:32]: Tests in a following PR: https://github.com/scality/cloudserver/pull/2609

**Review by miniscruff — APPROVED** [review]
LGTM

**Comment by dora-korpar** [comment]
@bert-e approve

**Comment by dora-korpar** [comment]
@bert-e reset

**Comment by dora-korpar** [comment]
@bert-e reset

**Pull request merged**